control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -0,0 +1,367 @@
+// owners.
+func (r *Controller) getWorkloadSelectorFromEndpoints(ctx context.Context, pf PodFetcher, endpoints *corev1.Endpoints) (*pbcatalog.WorkloadSelector, error) {
+	podPrefixes := make(map[string]any)
+	podExactNames := make(map[string]any)
+	var errs error
+	for address := range allAddresses(endpoints.Subsets) {
+		if address.TargetRef != nil && address.TargetRef.Kind == "Pod" {
+			podName := types.NamespacedName{Name: address.TargetRef.Name, Namespace: endpoints.Namespace}
+
+			// Accumulate owner prefixes and exact pod names for Consul workload selector.
+			// If this pod is already covered by a known owner prefix, skip it.
+			// If not, fetch the owner. If the owner has a unique prefix, add it to known prefixes.
+			// If not, add the pod name to exact name matches.
+			maybePodOwnerPrefix := getOwnerPrefixFromPodName(podName.Name)
+			if _, ok := podPrefixes[maybePodOwnerPrefix]; !ok {
+				pod, err := pf.GetPod(ctx, podName)
+				if err != nil {
+					r.Log.Error(err, "failed to get pod", "name", podName.Name, "ns", endpoints.Namespace)
+					errs = multierror.Append(errs, err)
+					continue
+				}
+				// Add to workload selector values.
+				// Pods can appear more than once in Endpoints subsets, so we use a set for exact names as well.
+				if prefix := getOwnerPrefixFromPod(pod); prefix != "" {
+					podPrefixes[prefix] = true
+				} else {
+					podExactNames[podName.Name] = true
+				}
+			}
+		}
+	}
+	return getWorkloadSelector(podPrefixes, podExactNames), errs
+}
+
+// allAddresses combines all Endpoints subset addresses to a single set. Service registration by this controller
+// operates independent of health, and an address can appear in multiple subsets if it has a mix of ready and not-ready
+// ports, so we combine them here to simplify iteration.
+func allAddresses(subsets []corev1.EndpointSubset) map[corev1.EndpointAddress]any {
+	m := make(map[corev1.EndpointAddress]any)
+	for _, sub := range subsets {
+		for _, readyAddress := range sub.Addresses {
+			m[readyAddress] = true
+		}
+		for _, notReadyAddress := range sub.NotReadyAddresses {
+			m[notReadyAddress] = true
+		}
+	}
+	return m
+}
+
+// getOwnerPrefixFromPodName extracts the owner name prefix from a pod name.
+func getOwnerPrefixFromPodName(podName string) string {
+	podNameParts := strings.Split(podName, "-")
+	return strings.Join(podNameParts[:len(podNameParts)-1], "-")
+}
+
+// getOwnerPrefixFromPod returns the common name prefix of the pod, if the pod is a member of a set with a unique name
+// prefix. Currently, this only applies to ReplicaSets.
+//
+// We have to fetch the owner and check its type because pod names cannot be disambiguated from pod owner names due to
+// the `-` delimiter and unique ID parts also being valid name components.
+//
+// If the pod owner does not have a unique name, the empty string is returned.
+func getOwnerPrefixFromPod(pod *corev1.Pod) string {
```

Comment:
Just an interesting fact, I noticed there is a field [generateName](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#objectmeta-v1-meta) that includes the shared prefixes that pods used. I think the current approach is more robust though.

Reply:
Ah yeah, thanks for pointing out! I think I ran across some doc referring to that field early on and had the same thought; it didn't seem guaranteed enough to hang implementation on over the owner ref.

For fun, here's an example:
```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2023-08-28T15:55:38Z"
  generateName: fluentd-elasticsearch-85996f8dbd-
  labels:
    app: nginx
    pod-template-hash: 85996f8dbd
  name: fluentd-elasticsearch-85996f8dbd-9r67q
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: fluentd-elasticsearch-85996f8dbd
    uid: 610a6961-944c-4863-9738-7d783dd1992c
  resourceVersion: "1841"
  uid: 28e1fdd5-0107-4ddc-890b-e07065d6000f
```